### PR TITLE
Implement .enumerate_support() for Bernoulli, Categorical distributions

### PR DESCRIPTION
**Pull request by fritzo**
## Why?
This method is used heavily in Pyro distributions for summing-out discrete variables.

## How?

The `.enumerate_support()` method returns a Tensor (or Variable) with all possible elementwise values, enumerated over the leftmost dimension. The result can thus be passed to `dist.log_prob()` as if it were a batch of samples. Note that `.enumerate_support()` does not enumerate over the cartesian product: doing so is exponentially expensive, rarely useful, and easy to accomplish efficiently via `itertools.product(dist.enumerate_support())`. In Pyro we originally implemented cartesian product but then switched to this more useful version.

This PR also includes the property `.has_enumerate_support` which is only true for discrete distributions.

Note this PR should be completely forwards-compatible with @neerajprad's upcoming PR to change batch shape https://github.com/probtorch/pytorch/issues/30.

### CODE REVIEW

**Comment by fritzo** [comment]
CC: @apaszke @neerajprad

File: torch/distributions.py [hunk 1]
```diff
@@ -77,6 +78,20 @@ def log_prob(self, value):
         """
         raise NotImplementedError
 
+    def enumerate_support(self):
+        """
+        Returns tensor containing all values supported by a discrete
+        distribution. The result will enumerate over dimension 0.
+
+        Note that this enumerates over all batched variables in lock-step
+        ``[[0, 0], [1, 1], ...]``. To iterate over the full Cartesian product
```
Comment (apaszke) [inline torch/distributions.py]: Is this always equivalent to taking a flat support vector, and expanding it to have as many columns as batch size?

File: torch/distributions.py [hunk 2]
```diff
@@ -163,6 +191,17 @@ def log_prob(self, value):
 
         return p.gather(-1, value.unsqueeze(-1)).squeeze(-1).log()
 
+    def enumerate_support(self):
+        batch_shape, event_size = self.probs.shape[:-1], self.probs.shape[-1]
+        values = torch.arange(event_size).long()
+        values = values.view((-1,) + (1,) * len(batch_shape))
+        values = values.expand((-1,) + batch_shape)
+        if self.probs.is_cuda:
+            values = values.cuda()
```
Comment (apaszke) [inline torch/distributions.py]: you probably want to retain the device: `values = values.cuda(self.probs.get_device())`

**Review by apaszke — COMMENTED** [review]
Looks good. I think it needs a tiny change to retain the device correctly.

**Comment by neerajprad** [comment]
You'll need to merge in the changes from #4131.

File: torch/distributions.py (hunk 2 above)
Comment (fritzo) [reply torch/distributions.py]: Thanks! I'll also make changes in Pyro.

File: torch/distributions.py (hunk 1 above)
Comment (fritzo) [reply torch/distributions.py]: Correct. The final shape should be `(cardinality,) + batch_shape`. In multivariate distributions like Multinomial or OneHotCategorical, the shape will be `(cardinality,) + batch_shape + event_shape`, but for the simple distributions in this PR, `event_shape = ()`.

**Comment by neerajprad** [comment]
Looks good to me.

**Comment by fritzo** [comment]
OK, I've addressed all review comments.

**Comment by apaszke** [comment]
It's failing the lint 😕

File: torch/distributions.py (hunk 1 above)
Comment (fritzo) [reply torch/distributions.py]: Done, thanks for the suggestion.

**Pull request merged**